In [9]:
# 필요한 라이브러리 임포트
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [10]:
import numpy as np
import pandas as pd
from sklearn.datasets import fetch_openml

#데이터 로드
mnist = fetch_openml('mnist_784', version=1)
X, y = mnist.data, mnist.target

In [13]:
X_train, X_test, y_train, y_test = X[:60000], X[60000:], y[:60000], y[60000:]

In [14]:
import numpy as np
from sklearn.neighbors import KNeighborsClassifier

MNIST 데이터셋으로 분류기를 만들어 테스트 세트에서 97% 정확도를 달성해보세요.

In [17]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score


# K-최근접 이웃 분류기 생성
# Class sklearn.neighbors.KNeighborsClassifier(n_neighbors=5, *, weights='uniform',algorithm='auto', leaf_size=30, p=2, metric='minkowski', metric_params=None, n_jobs=None)
knn = KNeighborsClassifier(n_neighbors = 3, weights='distance')

# 학습
knn.fit(X_train, y_train)

# 예측
y_pred = knn.predict(X_test)

# 정확도 평가
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")

Accuracy: 0.9717


Accuracy: 0.9707 # n_neighbors = 3
Accuracy: 0.9717 # n_neighbors = 3, weights='distance'

# hyperparameter tunning 

In [20]:
from sklearn.model_selection import GridSearchCV


In [21]:
# KNN 분류 모델을 사용한다고 가정
model = KNeighborsClassifier()

# 탐색할 하이퍼파라미터의 그리드 설정
param_grid = {
    'n_neighbors': [3, 4, 5, 6, 7],  # 이웃의 수
    'weights': ['uniform', 'distance'],  # 가중치 방식
    'metric': ['euclidean', 'manhattan']  # 거리 계산 방식
}


In [22]:
# GridSearchCV 설정: 모델, 파라미터 그리드, 교차검증 횟수(cv)
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5, scoring='accuracy')


In [23]:
grid_search.fit(X_train, y_train)

KeyboardInterrupt: 

In [ ]:
print("Best Parameters:", grid_search.best_params_)
print("Best Score:", grid_search.best_score_)

In [25]:
import nbformat as nbf

# 변환할 파이썬 스크립트 파일 경로
python_script = '/Users/apple/Desktop/Python/Smarcle/MachineLearning/제출.py'

# 빈 Jupyter 노트북 객체 생성
notebook = nbf.v4.new_notebook()

# 파이썬 스크립트 파일 열기
with open(python_script, 'r', encoding='utf-8') as f:
    code = f.read()

# 코드를 Jupyter 노트북의 코드 셀로 변환
notebook.cells.append(nbf.v4.new_code_cell(code))

# 변환된 파일을 .ipynb 파일로 저장
with open('/Users/apple/Desktop/Python/Smarcle/MachineLearning/pair.ipynb', 'w', encoding='utf-8') as f:
    nbf.write(notebook, f)

print("변환 완료: converted_notebook.ipynb 파일이 생성되었습니다.")

변환 완료: converted_notebook.ipynb 파일이 생성되었습니다.
